### DISCUSSION TITLE
WebSocketActor GlobalActor in CaseStudies

### DISCUSSION THREAD

**Discussion by JackYoustra** [pointfreeco/swift-composable-architecture#1536] [Q&A]
I'm looking for a way for a dependency to synchronize its entrypoints, and found the useful `WebSocketClient` code. I see that `WebSocketActor` conforms to `GlobalActor` but I can't figure out why. Can anyone help?

EDIT:
A few other questions as I look at the code more:

- Why a shared singleton for the `WebSocketActor`? It seems like only one will ever be created (it's wrapped in the synchronized constructor for `liveValue`).
- The semantics of `open` and `receive:` seem strange: you have a signature of, in receive's case, `@Sendable (Any.Type) async throws -> AsyncStream<TaskResult<Message>>`. Is the actor barrier the reason why you can't have a synchronous method returning an `AsyncStream`? Any reason why there can't be an adapter written for this? Am I completely off-base?

**Comment by mbrandonw** [comment]
This example is maybe a little more complex than is really necessary. The web socket client supports servicing any number of web socket connections, hence the `id` being passed around everywhere. So, there is a single global actor for synchronizing access to a dictionary mapping ids to socket connections.

We don't actually use that power in the case study since there is only one single connection. We could maybe simplify. Stephen and I will discuss.

**Reply by JackYoustra** [reply]
(or complexity so you do use it, I found it illuminating!)
